repo.go

```diff
@@ -37,12 +37,138 @@ type Interface interface {
 // Charm store references will use the provided parameters.
 // Local references will use the provided path.
 func InferRepository(ref *charm.URL, charmStoreParams NewCharmStoreParams, localRepoPath string) (Interface, error) {
-	switch ref.Schema {
+	dis := NewDispatcher(charmStoreParams, localRepoPath)
+	inferred, err := dis.Infer(ref.Schema)
+	if err != nil {
+		return nil, err
+	}
+	repo, err := inferred.NewRepo()
+	if err != nil {
+		return nil, err
+	}
+	return repo, nil
+}
+
+// Dispatcher supports selecting and performing repo-related behavior
+// based on charm schemas. Behavior selection is independent of the
+// source of the schema (e.g. charm.URL).
+//
+// Recognized schemas are:
+//   cs
+//   local
+type Dispatcher struct {
+	// Factory is the charm repo factory that the dispatcher uses.
+	Factory Factory
+
+	// Handlers maps charm schemas to functions that will be called
+	// in response to the corresponding schema. A missing entry
+	// (schema not found in the map) indicates that the schema is not
+	// supported. If the entry is there but holds a nil function
+	// then that indicates a no-op.
+	Handlers map[string]func(schema string, repo Interface) error
+}
+
+// NewDispatcher returns a new repo dispatcher with no-op handlers for
+// each of the recognized schemas. The dispatcher's repo factory is
+// created using the provided charm store and local repo info.
+func NewDispatcher(charmStoreParams NewCharmStoreParams, localRepoPath string) *Dispatcher {
+	return &Dispatcher{
+		Factory:  NewFactory(charmStoreParams, localRepoPath),
+		Handlers: map[string]func(string, Interface) error{"cs": nil, "local": nil},
+	}
+}
+
+// Infer determines the repo and handler to use for the given schema.
+// It does not execute any behavior relative to the repo.
+func (dis Dispatcher) Infer(schema string) (*Inferred, error) {
```

Comment:
This isn't about testing, but rather about using InferRepository() with an existing cilent.  It was also an attempt to extract the mapping of schema to repo factory func in a portable way.  I agree that the code is simpler if you drop the former.  Regardless, I'll drop most of this code from the patch.